drivers/i2c/i2c_ite_enhance.c

```diff
@@ -568,6 +605,311 @@ static int i2c_enhance_transfer(const struct device *dev, struct i2c_msg *msgs,
+	for (i = 0; i < data->cq_msgs[0].len; i++) {
+		i2c_cq_pckt->wdata[i] = data->cq_msgs[0].buf[i];
+	}
+
+	/* Set commands in RAM. (command byte for read) */
+	num_bit_2_0 = (data->cq_msgs[1].len - 1) & I2C_CQ_CMD_L_NUM_BIT_2_0;
+	num_bit_10_3 = ((data->cq_msgs[1].len - 1) >> 3) & 0xff;
+	i2c_cq_pckt->wdata[i++] = I2C_CQ_CMD_L_RS | I2C_CQ_CMD_L_RW |
+				I2C_CQ_CMD_L_P | I2C_CQ_CMD_L_E | num_bit_2_0;
+	i2c_cq_pckt->wdata[i] = num_bit_10_3;
+}
+
+static int enhanced_i2c_cmd_queue_trans(const struct device *dev)
+{
+	struct i2c_enhance_data *data = dev->data;
+	const struct i2c_enhance_config *config = dev->config;
+	uint8_t *base = config->base;
+
+	/* ISR of command queue mode. */
+	if (k_is_in_isr()) {
+		/* Device 1 finish IRQ. */
+		if (IT8XXX2_I2C_FST(base) & IT8XXX2_I2C_FST_DEV1_IRQ) {
+			uint8_t msgs_idx = data->num_msgs - 1;
+
+			/* Get data if this is a read transaction. */
+			for (int i = 0; i < data->cq_msgs[msgs_idx].len; i++) {
+				data->cq_msgs[msgs_idx].buf[i] =
+				data->i2c_cq_mode_rx_dlm[i];
+			}
+		} else {
+			/* Device 1 error have occurred. eg. nack, timeout... */
+			if (IT8XXX2_I2C_NST(base) & IT8XXX2_I2C_NST_ID_NACK) {
+				data->err = E_HOSTA_ACK;
+			} else {
+				data->err = IT8XXX2_I2C_STR(base) &
+						E_HOSTA_ANY_ERROR;
+			}
+		}
+		/* Reset bus. */
+		IT8XXX2_I2C_CTR(base) = E_STS_AND_HW_RST;
+		IT8XXX2_I2C_CTR1(base) = 0;
+
+		return 0;
+	}
+
+	/* State reset and hardware reset. */
+	IT8XXX2_I2C_CTR(base) = E_STS_AND_HW_RST;
+	/* Set "PSR" registers to decide the i2c speed. */
+	i2c_enhanced_port_set_frequency(dev, config->bitrate);
+	/* Set time out register. port D, E, or F clock/data low timeout. */
+	IT8XXX2_I2C_TOR(base) = I2C_CLK_LOW_TIMEOUT;
+
+	if (data->num_msgs == 2) {
+		/* I2C write to read of command queue mode. */
+		enhanced_i2c_cq_write_to_read(dev);
+	} else {
+		/* I2C read of command queue mode. */
+		if (data->cq_msgs[0].flags & I2C_MSG_READ) {
+			enhanced_i2c_cq_read(dev);
+		/* I2C write of command queue mode. */
+		} else {
+			enhanced_i2c_cq_write(dev);
+		}
+	}
+
+	/* Enable i2c module with command queue mode. */
+	IT8XXX2_I2C_CTR1(base) = IT8XXX2_I2C_MDL_EN | IT8XXX2_I2C_COMQ_EN;
+	/* One shot on device 1. */
+	IT8XXX2_I2C_MODE_SEL(base) = 0;
+	IT8XXX2_I2C_CTR2(base) = 1;
+	/* Start */
+	i2c_enhance_disable_cpu_idle(config->port);
+	IT8XXX2_I2C_CTR(base) = E_START_CQ;
+
+	return 1;
+}
+
+static int i2c_enhance_cq_transfer(const struct device *dev,
+				   struct i2c_msg *msgs)
+{
+	struct i2c_enhance_data *data = dev->data;
+	const struct i2c_enhance_config *config = dev->config;
+	int res = 0;
+
+	data->err = 0;
+	data->cq_msgs = msgs;
+
+	/* Make sure we're in a good state to start */
+	if (i2c_bus_not_available(dev)) {
+		/* Recovery I2C bus */
+		i2c_recover_bus(dev);
+		/*
+		 * After resetting I2C bus, if I2C bus is not available
+		 * (No external pull-up), drop the transaction.
+		 */
+		if (i2c_bus_not_available(dev)) {
+			return -EIO;
+		}
+	}
```

Comment:
Can this check get moved to `i2c_enhance_raw_transfer()`?

Reply:
Done.